apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -976,8 +1151,51 @@ fn handle_error<T: Into<opentelemetry::global::Error>>(err: T) {
     }
 }
 
+pub(crate) fn is_span_sampled(context: &Context) -> bool {
+    Span::current().context().span().span_context().is_sampled()
+        && !context
+            .get(FTV1_DO_NOT_SAMPLE)
+            .unwrap_or_default()
+            .unwrap_or(false)
+}
+
 register_plugin!("apollo", "telemetry", Telemetry);
 
+/// This enum is a partial cleanup of the telemetry plugin logic.
+///
+#[derive(Copy, Clone)]
+enum ApolloFtv1Handler {
+    Enabled,
+    Disabled,
+}
+
+impl ApolloFtv1Handler {
+    fn request_ftv1(&self, mut req: SubgraphRequest) -> SubgraphRequest {
+        if let ApolloFtv1Handler::Enabled = self {
+            if is_span_sampled(&req.context) {
+                req.subgraph_request.headers_mut().insert(
+                    "apollo-federation-include-trace",
+                    HeaderValue::from_static("ftv1"),
+                );
+            }
+        }
+        req
+    }
+
+    fn store_ftv1(&self, resp: SubgraphResponse) -> SubgraphResponse {
+        // Stash the FTV1 data
+        if let ApolloFtv1Handler::Enabled = self {
```

Comment:
is this if let here for performance reasons? i suspect there wont be any ftv1 extension in the response if we havent added `apollo-federation-include-trace` in the request

Reply:
yep